### PR TITLE
[InstCombine] Fold `fcmp pred sqrt(X), 0.0 -> fcmp pred2 X, 0.0`

### PR DESCRIPTION
**Pull request by dtcxzyw**
Proof (Please run alive-tv with larger smt-to): https://alive2.llvm.org/ce/z/-aqixk
FMF propagation: https://alive2.llvm.org/ce/z/zyKK_p

```
sqrt(X) < 0.0 --> false
sqrt(X) u>= 0.0 --> true
sqrt(X) u< 0.0 --> X u< 0.0
sqrt(X) u<= 0.0 --> X u<= 0.0
sqrt(X) > 0.0 --> X > 0.0
sqrt(X) >= 0.0 --> X >= 0.0
sqrt(X) == 0.0 --> X == 0.0
sqrt(X) u!= 0.0 --> X u!= 0.0
sqrt(X) <= 0.0 --> X == 0.0
sqrt(X) u> 0.0 --> X u!= 0.0
sqrt(X) u== 0.0 --> X u<= 0.0
sqrt(X) != 0.0 --> X > 0.0
!isnan(sqrt(X)) --> X >= 0.0
isnan(sqrt(X)) --> X u< 0.0
```

In most cases, `sqrt` cannot be eliminated since it has multiple uses. But this patch will break data dependencies and allow optimizer to sink expensive `sqrt` calls into successor blocks.

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/fcmp.ll [hunk 1]
```diff
@@ -2117,3 +2117,175 @@ define <8 x i1> @fcmp_ogt_fsub_const_vec_denormal_preserve-sign(<8 x float> %x,
   %cmp = fcmp ogt <8 x float> %fs, zeroinitializer
   ret <8 x i1> %cmp
 }
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fcmp.ll:2120]: Test some compares with -0 too

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -7980,6 +7980,63 @@ static Instruction *foldFabsWithFcmpZero(FCmpInst &I, InstCombinerImpl &IC) {
   }
 }
 
+/// Optimize sqrt(X) compared with zero.
+static Instruction *foldSqrtWithFcmpZero(FCmpInst &I, InstCombinerImpl &IC) {
+  Value *X;
+  if (!match(I.getOperand(0), m_Sqrt(m_Value(X))))
+    return nullptr;
+
+  if (!match(I.getOperand(1), m_PosZeroFP()))
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7989]: This should work for anyZeroFP

Comment (goldsteinn) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7989]: Most: https://alive2.llvm.org/ce/z/nwWqUB

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:7989]: fcmp with -0.0 is not canonical:

https://github.com/llvm/llvm-project/blob/10df3207434e603be5f7e9b3036d821dd5623d3a/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L8135-L8138

File: llvm/test/Transforms/InstCombine/fcmp.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/fcmp.ll:2120]: Should still test with -0 , canonical or not

File: llvm/test/Transforms/InstCombine/fcmp.ll [hunk 3]
```diff
@@ -2266,6 +2296,17 @@ define i1 @fcmp_sqrt_zero_uno(half %x) {
   ret i1 %cmp
 }
 
+; Make sure that ninf is cleared.
+define i1 @fcmp_sqrt_zero_uno_fmf(half %x) {
+; CHECK-LABEL: @fcmp_sqrt_zero_uno_fmf(
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp ult half [[X:%.*]], 0xH0000
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %sqrt = call half @llvm.sqrt(half %x)
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fcmp.ll:2305]: Missing type suffixes on all the sqrt calls.

Also, can you keep ninf if the sqrt is also ninf?

Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fcmp.ll]: Done.

**Review by arsenm — APPROVED** [review]

**Pull request merged**